File: cpp/Mod_flood.cpp
#pragma once

String flood(String s, int width, int x, int y){
	List<Char> str = s2lc(s);
	int pos = width*y+x;
	Char color = get(str, pos);
	Char newColor;
	if(color == 'b')
		newColor = 'w';
	else 
		newColor = 'b';
	
	str = set(str, pos, newColor);
	
    //left neighbor
	if(x != 0){
		int newX = x-1;
		int newPos = width*y+newX;
		if(get(str, newPos) != newColor){
			str = s2lc(flood(lc2s(str), width, newX, y));
		}
	}
	
    //bottom neighbor
	if(y != 0){
		int newY = y-1;
		int newPos = width*newY+x;
		if(get(str, newPos) != newColor){
			str = s2lc(flood(lc2s(str), width, x, newY));
		}
	}
	
    //right neighbor
	if(x != (width-1)){
		int newX = x+1;
		int newPos = width*y+newX;
		if(get(str, newPos) != newColor){
			str = s2lc(flood(lc2s(str), width, newX, y));
		}
	}
	
    //top neighbor
	if(y != sp_div(length(str), width)-1) {
		int newY = y+1;
		int newPos = width*newY+x;
		if(get(str, newPos) != newColor){
			str = s2lc(flood(lc2s(str), width, x, newY));
		}
	}
	
	return lc2s(str);
}

File: hs/Mod_flood.hs
module Mod_flood ( flood ) where
import Prelude hiding (and, or) 
import Debug.Trace
import Splib
import Types

flood :: String -> Int -> Int -> Int -> String
flood str width x y = floodHelper (floodHelper (floodHelper (floodHelper (set str pos color) width x (y+1) color) width x (y-1) color) width (x+1) y color) width (x-1) y color
                where 
                 pos = width*(y)+(x+1)-1
                 color | get str pos == 'b' = 'w'
                       | otherwise = 'b'

floodHelper :: String -> Int -> Int -> Int -> Char -> String
floodHelper str width x y newColor 
 | get str safePos == newColor = str
 | otherwise = flood str width x y
        where 
          pos = width*(y)+(x+1)-1
          safeX = sp_max (sp_min x (width-1)) 0
          safeY = sp_max (sp_min y ((sp_div (length str) width)-1)) 0
          safePos = width*safeY+safeX